prog/minio/setup_minio.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+class Prog::Minio::SetupMinio < Prog::Base
+  subject_is :minio_server
+
+  label def install_minio
+    case minio_server.vm.sshable.cmd("common/bin/daemonizer --check install_minio")
+    when "Succeeded"
+      pop "minio is installed"
+    when "Failed", "NotStarted"
+      minio_server.vm.sshable.cmd("common/bin/daemonizer 'minio/bin/install_minio #{Config.minio_version}' install_minio")
+    end
+    nap 5
+  end
+
+  label def configure_minio
+    def_minio = <<ECHO
+MINIO_VOLUMES="#{minio_server.minio_volumes}"
+MINIO_OPTS="--console-address :9001"
+MINIO_ROOT_USER="#{minio_server.cluster.admin_user}"
+MINIO_ROOT_PASSWORD="#{minio_server.cluster.admin_password}"
+ECHO
+
+    hosts = <<ECHO
+#{minio_server.cluster.generate_etc_hosts_entry}
+ECHO
+
+    minio_server.vm.sshable.cmd("sudo tee -a /etc/default/minio", stdin: def_minio)
+    minio_server.vm.sshable.cmd("sudo tee -a /etc/hosts", stdin: hosts)
+    minio_server.vm.sshable.cmd("sudo chown -R minio-user:minio-user /etc/default/minio")
+
+    pop "minio is configured"
+  end
+
+  label def mount_data_disks
+    case minio_server.vm.sshable.cmd("common/bin/daemonizer --check format_disks")
+    when "Succeeded"
+      minio_server.vm.sshable.cmd("sudo mkdir -p /minio")
+      minio_server.vm.vm_storage_volumes_dataset.order_by(:disk_index).where(Sequel[:vm_storage_volume][:boot] => false).all.each_with_index do |volume, i|
+        minio_server.vm.sshable.cmd("sudo mkdir -p /minio/dat#{i + 1}")
+        device_path = volume.device_path.shellescape
+        minio_server.vm.sshable.cmd("sudo common/bin/add_to_fstab #{device_path} /minio/dat#{i + 1} xfs defaults 0 0")
+        minio_server.vm.sshable.cmd("sudo mount #{device_path} /minio/dat#{i + 1}")
+      end
+      minio_server.vm.sshable.cmd("sudo chown -R minio-user:minio-user /minio")
+
+      pop "data disks are mounted"
+    when "Failed", "NotStarted"
+      cmd = minio_server.vm.vm_storage_volumes_dataset.order_by(:disk_index).where(Sequel[:vm_storage_volume][:boot] => false).all.map do |volume|
+        device_path = volume.device_path.shellescape
+        "sudo mkfs --type xfs #{device_path}"
+      end.join(" & ")
```

Comment:
If you join commands with & you make all commands background jobs except the last one. In this model you wouldn't get exceptions for failures (except from the last one), thus if one of the intermediate command fail, you would exit from this label as Succeeded.

Try this for example;
```ruby
vm.sshable.cmd("xxx") # throws exception

vm.sshable.cmd("sleep 1 & xxx") # throws exception

vm.sshable.cmd("xxx & sleep 1") # does not throw exception
```

You can use `&&`, but that would be slower as it would format the disks sequentially. Normally you can run "fg" to bring background jobs to foreground (thus getting the exit status), but that wouldn't work with daemonizer, also it is not reliable (other background jobs would interfere the result of fg).

I think it is OK to be a bit slower in this phase, it can be optimized if needed. Or you might find some bash tricks to make it work.

Reply:
This is a good catch. Thanks! I don't mind this being slow since this specific step is parallelized with installing minio which already consumes much more time than mount_disk. I don't think it will ever impact. I changed it to `&&`